horovod/common/operations.cc

```diff
@@ -865,6 +899,21 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
 
         ACTIVITY_END_ALL(entries, timeline)
       }
+#elif HOROVOD_GPU_ALLREDUCE == 'D'
+      if (!horovod_global.ddl_initialized) {
+        // Initialize DDL
+        auto ddl_options = std::getenv("DDL_OPTIONS");
+        if (ddl_options == nullptr) {
+          throw std::logic_error("DDL_OPTIONS env variable needs "
+                                 "to be set to use DDL.");
+        }
+        DDL_CHECK(entries, "ddl_init", ddl_init(ddl_options))
+        horovod_global.ddl_initialized = true;
+        horovod_global.ddl_local_device_id = first_entry.device;
+      } else if (horovod_global.ddl_local_device_id != first_entry.device) {
+        throw std::logic_error("DDL does not support more than one GPU device per process.");
```

Comment:
Instead of exception (on a background thread), can this do something like what DDL_CHECK does:
```
for (auto& e : (entries)) {
        timeline.End(e.tensor_name, nullptr);                                  \
        e.callback(Status::UnknownError(std::string(op_name) + " failed."));   \
}
return
```